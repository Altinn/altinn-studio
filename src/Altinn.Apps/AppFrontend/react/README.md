## Running the tests

### Lint checks

1. Navigate to the folder `src/Altinn.Apps/AppFrontend/react`.
2. Execute `npm ci`. This step is only nescessary if you have not already done it, or if you change branches.
3. Execute `npm run install-deps`. This step is only nescessary if you have not already done it, or if you change branches.
4. Execute `npm run lint`.

### Unit tests

1. Navigate to the folder `src/Altinn.Apps/AppFrontend/react`.
2. Execute `npm ci`. This step is only nescessary if you have not already done it, or if you change branches.
3. Execute `npm run install-deps`. This step is only nescessary if you have not already done it, or if you change branches.
4. Execute `npm run test`.
