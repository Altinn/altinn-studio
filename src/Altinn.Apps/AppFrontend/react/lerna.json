{
  "command": {
    "bootstrap": {
      "hoist": true
    }
  },
  "packages": [
    "./*"
  ],
  "version": "independent"
}
