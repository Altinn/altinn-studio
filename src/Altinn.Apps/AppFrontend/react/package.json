{
	"name": "app-frontend",
	"private": true,
	"scripts": {
		"install-deps": "lerna bootstrap --hoist --ci",
		"no-ci-hoist": "lerna bootstrap --hoist --no-ci",
		"install-deps-with-lock": "lerna bootstrap",
		"test": "lerna exec --concurrency 1 -- npm test",
		"install-all": "lerna exec -- npm ci",
		"clean-lock": "lerna exec -- rm ./package-lock.json",
		"clean-node-modules": "rm -rf ./node_modules && lerna exec --concurrency 1 -- rm -rf ./node_modules",
		"update-packages": "lerna exec --concurrency 1 -- npm update --dev",
		"list-outdated": "lerna run list-outdated"
	},
	"devDependencies": {
		"@typescript-eslint/eslint-plugin": "~4.8.2",
		"@typescript-eslint/parser": "~4.8.2",
		"eslint": "~7.14.0",
		"eslint-config-airbnb": "~18.2.1",
		"eslint-plugin-import": "~2.22.1",
		"eslint-plugin-jsx-a11y": "~6.4.1",
		"eslint-plugin-react": "~7.21.5",
		"eslint-plugin-react-hooks": "^4.2.0",
		"ini": "1.3.5",
		"lerna": "3.22.1"
	}
}
