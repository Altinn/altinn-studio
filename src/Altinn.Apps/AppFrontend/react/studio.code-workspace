{
	"folders": [
		{
			"path": "."
		}
  ],
  "settings": {
    "typescript.surveys.enabled": false,
    "typescript.tsdk": "node_modules/typescript/lib",
    "eslint.alwaysShowStatus": true,
    "eslint.validate": [
      "javascript",
      "javascriptreact",
      { "language": "typescript", "autoFix": true },
      { "language": "typescriptreact", "autoFix": true}
    ]
  }
}
