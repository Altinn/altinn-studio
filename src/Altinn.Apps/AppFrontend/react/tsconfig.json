{
  "compilerOptions": {
    "outDir": "compiled",
    "module": "commonjs",
    "target": "es5",
    "lib": [
      "es6",
      "dom",
      "es7",
      "esnext"
    ],
    "sourceMap": true,
    "allowJs": true,
    "jsx": "react",
    "moduleResolution": "node",
    "rootDir": "./",
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noImplicitAny": true,
    "strictNullChecks": false,
    "noUnusedLocals": true
  },
  "include": [
    "**/*.ts",
    "**/*.tsx",
  ],
  "exclude": [
    "node_modules",
    "build",
    "scripts",
    "acceptance-tests",
    "webpack",
    "jest",
    "setupTests.ts",
    "templates",
    "**/setupTests.ts"
  ],
}
