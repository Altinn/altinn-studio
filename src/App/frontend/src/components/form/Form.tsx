import React, { useEffect, useMemo } from 'react';
import { useLocation, useNavigate, useSearchParams } from 'react-router-dom';

import { Flex } from 'src/app-components/Flex/Flex';
import classes from 'src/components/form/Form.module.css';
import { MessageBanner } from 'src/components/form/MessageBanner';
import { ErrorReport, ErrorReportList } from 'src/components/message/ErrorReport';
import { ReadyForPrint } from 'src/components/ReadyForPrint';
import { NavigateToStartUrl } from 'src/components/wrappers/ProcessWrapper';
import { SearchParams } from 'src/core/routing/types';
import { useAppName, useAppOwner } from 'src/core/texts/appTexts';
import { useApplicationMetadata } from 'src/features/applicationMetadata/ApplicationMetadataProvider';
import { useAllAttachments } from 'src/features/attachments/hooks';
import { FileScanResults } from 'src/features/attachments/types';
import { useExpandedWidthLayouts, useLayoutLookups } from 'src/features/form/layout/LayoutsContext';
import { useUiConfigContext } from 'src/features/form/layout/UiConfigContext';
import { usePageSettings } from 'src/features/form/layoutSettings/LayoutSettingsContext';
import { useLaxInstanceId } from 'src/features/instance/InstanceContext';
import { useLanguage } from 'src/features/language/useLanguage';
import { useOnFormSubmitValidation } from 'src/features/validation/callbacks/onFormSubmitValidation';
import { useTaskErrors } from 'src/features/validation/selectors/taskErrors';
import { useQueryKey } from 'src/hooks/navigation';
import { useAsRef } from 'src/hooks/useAsRef';
import { useCurrentView, useNavigatePage } from 'src/hooks/useNavigatePage';
import { getComponentCapabilities } from 'src/layout';
import { GenericComponent } from 'src/layout/GenericComponent';
import { getPageTitle } from 'src/utils/getPageTitle';
import { NodesInternal } from 'src/utils/layout/NodesContext';
import type { AnyValidation, BaseValidation, NodeRefValidation } from 'src/features/validation';

interface FormState {
  hasRequired: boolean;
  mainIds: string[];
  errorReportIds: string[];
  formErrors: NodeRefValidation<AnyValidation<'error'>>[];
  taskErrors: BaseValidation<'error'>[];
}

export function Form() {
  const currentPageId = useCurrentView();

  return <FormPage currentPageId={currentPageId} />;
}

export function FormPage({ currentPageId }: { currentPageId: string | undefined }) {
  const [searchParams, setSearchParams] = useSearchParams();
  const shouldValidateFormPage = searchParams.get(SearchParams.Validate);
  const onFormSubmitValidation = useOnFormSubmitValidation();

  useEffect(() => {
    if (shouldValidateFormPage) {
      onFormSubmitValidation();
      setSearchParams((params) => {
        params.delete(SearchParams.Validate);
        return searchParams;
      });
    }
  }, [onFormSubmitValidation, searchParams, setSearchParams, shouldValidateFormPage]);

  const { isValidPageId } = useNavigatePage();
  const appName = useAppName();
  const appOwner = useAppOwner();
  const { langAsString } = useLanguage();
  const { hasRequired, mainIds, errorReportIds, formErrors, taskErrors } = useFormState(currentPageId);
  const requiredFieldsMissing = NodesInternal.usePageHasVisibleRequiredValidations(currentPageId);
  const allAttachments = useAllAttachments();

  const hasInfectedFiles = Object.values(allAttachments || {}).some((attachments) =>
    (attachments || []).some(
      (attachment) => attachment.uploaded && attachment.data.fileScanResult === FileScanResults.Infected,
    ),
  );

  useRedirectToStoredPage();
  useSetExpandedWidth();

  if (!currentPageId || !isValidPageId(currentPageId)) {
    return <NavigateToStartUrl forceCurrentTask={false} />;
  }

  const hasSetCurrentPageId = langAsString(currentPageId) !== currentPageId;

  if (!hasSetCurrentPageId) {
    window.logWarnOnce(
      `You have not set a page title for this page. This is highly recommended for user experience and WCAG compliance and will be required in the future.
       To add a title to this page, add this to your language resource file (for example language.nb.json):

      {
        "id": "${currentPageId}",
         "value": "Your custom title goes here"
      }`,
    );
  }

  return (
    <>
      <title>{`${getPageTitle(appName, hasSetCurrentPageId ? langAsString(currentPageId) : undefined, appOwner)}`}</title>
      {hasRequired && (
        <MessageBanner
          error={requiredFieldsMissing}
          messageKey='form_filler.required_description'
        />
      )}
      <Flex
        container
        spacing={6}
        alignItems='flex-start'
      >
        {mainIds.map((id) => (
          <GenericComponent
            key={id}
            baseComponentId={id}
          />
        ))}
        <Flex
          item={true}
          size={{ xs: 12 }}
          aria-live='polite'
          className={classes.errorReport}
        >
          <ErrorReport
            show={formErrors.length > 0 || taskErrors.length > 0 || hasInfectedFiles}
            errors={
              <ErrorReportList
                formErrors={formErrors}
                taskErrors={taskErrors}
              />
            }
          >
            {errorReportIds.map((id) => (
              <GenericComponent
                key={id}
                baseComponentId={id}
              />
            ))}
          </ErrorReport>
        </Flex>
      </Flex>
      <ReadyForPrint type='load' />
      <HandleNavigationFocusComponent />
    </>
  );
}

/**
 * Redirects users that had a stored page in their local storage to the correct
 * page, and later removes this currentViewCacheKey from localstorage, as
 * it is no longer needed.
 */
function useRedirectToStoredPage() {
  const pageKey = useCurrentView();
  const { isValidPageId, navigateToPage } = useNavigatePage();
  const applicationMetadataId = useApplicationMetadata()?.id;

  const instanceId = useLaxInstanceId();
  const currentViewCacheKey = instanceId ?? applicationMetadataId;

  useEffect(() => {
    if (!pageKey && !!currentViewCacheKey) {
      const lastVisitedPage = localStorage.getItem(currentViewCacheKey);
      if (lastVisitedPage !== null && isValidPageId(lastVisitedPage)) {
        localStorage.removeItem(currentViewCacheKey);
        navigateToPage(lastVisitedPage, { replace: true });
      }
    }
  }, [pageKey, currentViewCacheKey, isValidPageId, navigateToPage]);
}

/**
 * Sets the expanded width for the current page if it is defined in the currently viewed layout-page
 */
function useSetExpandedWidth() {
  const currentPageId = useCurrentView();
  const expandedPagesFromLayout = useExpandedWidthLayouts();
  const expandedWidthFromSettings = usePageSettings().expandedWidth;
  const { setExpandedWidth } = useUiConfigContext();

  useEffect(() => {
    let defaultExpandedWidth = false;
    if (currentPageId && expandedPagesFromLayout[currentPageId] !== undefined) {
      defaultExpandedWidth = !!expandedPagesFromLayout[currentPageId];
    } else if (expandedWidthFromSettings !== undefined) {
      defaultExpandedWidth = expandedWidthFromSettings;
    }
    setExpandedWidth(defaultExpandedWidth);
  }, [currentPageId, expandedPagesFromLayout, expandedWidthFromSettings, setExpandedWidth]);
}

const emptyArray = [];
function useFormState(currentPageId: string | undefined): FormState {
  const lookups = useLayoutLookups();
  const topLevelIds = currentPageId ? (lookups.topLevelComponents[currentPageId] ?? emptyArray) : emptyArray;
  const { formErrors, taskErrors } = useTaskErrors();
  const hasErrors = Boolean(formErrors.length) || Boolean(taskErrors.length);

  const [mainIds, errorReportIds] = useMemo(() => {
    if (!hasErrors) {
      return [topLevelIds, emptyArray];
    }

    const toMainLayout: string[] = [];
    const toErrorReport: string[] = [];

    for (const id of [...topLevelIds].reverse()) {
      const type = lookups.allComponents[id]?.type;
      if (!type) {
        continue;
      }

      const capabilities = getComponentCapabilities(type);
      const isButtonLike = type === 'ButtonGroup' || (capabilities.renderInButtonGroup && type !== 'Custom');
      if (isButtonLike && toMainLayout.length === 0) {
        toErrorReport.push(id);
      } else {
        toMainLayout.push(id);
      }
    }

    return [toMainLayout.reverse(), toErrorReport.reverse()];
  }, [hasErrors, lookups.allComponents, topLevelIds]);

  const hasRequired =
    (currentPageId &&
      lookups.allPerPage[currentPageId]?.some((id) => {
        const layout = lookups.allComponents[id];
        return layout && 'required' in layout && layout.required !== false;
      })) ||
    false;

  return {
    hasRequired,
    mainIds,
    errorReportIds,
    formErrors,
    taskErrors,
  };
}

function HandleNavigationFocusComponent() {
  const onFormSubmitValidation = useOnFormSubmitValidation();
  const exitSubform = useQueryKey(SearchParams.ExitSubform)?.toLocaleLowerCase() === 'true';
  const validate = useQueryKey(SearchParams.Validate)?.toLocaleLowerCase() === 'true';
  const navigate = useNavigate();
  const searchStringRef = useAsRef(useLocation().search);

  React.useEffect(() => {
    (async () => {
      // Replace URL if we have query params
      if (exitSubform || validate) {
        const location = new URLSearchParams(searchStringRef.current);
        location.delete(SearchParams.ExitSubform);
        const baseHash = window.location.hash.slice(1).split('?')[0];
        const nextLocation = location.size > 0 ? `${baseHash}?${location.toString()}` : baseHash;
        navigate(nextLocation, { replace: true });
      }
    })();
  }, [navigate, searchStringRef, exitSubform, validate, onFormSubmitValidation]);

  return null;
}
