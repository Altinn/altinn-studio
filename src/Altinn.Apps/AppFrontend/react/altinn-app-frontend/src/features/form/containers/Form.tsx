/* eslint-disable no-undef */
import Grid from '@material-ui/core/Grid';
import * as React from 'react';
import { useSelector } from 'react-redux';
import { IRuntimeState } from '../../../types';
import { ILayout, ILayoutComponent, ILayoutGroup } from '../layout';
import { GroupContainer } from './GroupContainer';
import { renderGenericComponent } from '../../../utils/layout';

export function Form() {
  const [filteredLayout, setFilteredLayout] = React.useState<any[]>([]);
  const [currentLayout, setCurrentLayout] = React.useState<string>();

  const currentView = useSelector((state: IRuntimeState) => state.formLayout.uiConfig.currentView);
  const layout: ILayout =
    useSelector((state: IRuntimeState) => state.formLayout.layouts[state.formLayout.uiConfig.currentView]);
  const hiddenComponents: string[] = useSelector((state: IRuntimeState) => state.formLayout.uiConfig.hiddenFields);

  React.useEffect(() => {
    setCurrentLayout(currentView);
  }, [currentView]);

  React.useEffect(() => {
    let componentsToRender: any[] = layout;
    let renderedInGroup: string[] = [];
    if (layout) {
      const groupComponents = layout.filter((component) => component.type.toLowerCase() === 'group');
      groupComponents.forEach((component: ILayoutGroup) => {
        renderedInGroup = renderedInGroup.concat(component.children);
      });
    }

    if (layout && hiddenComponents) {
      componentsToRender = layout.filter((component) => {
        return !hiddenComponents.includes(component.id) && !renderedInGroup.includes(component.id);
      });
    }
    setFilteredLayout(componentsToRender);
  }, [layout, hiddenComponents]);

  function RenderGenericComponent(component: ILayoutComponent) {
    return renderGenericComponent(component);
  }

  function RenderLayoutGroup(layoutGroup: ILayoutGroup): JSX.Element {
    const groupComponents = layoutGroup.children.map((child) => {
      return layout.find((c) => c.id === child) as ILayoutComponent;
    });
    const repeating = layoutGroup.maxCount > 1;
    if (!repeating) {
      // If not repeating, treat as regular components
      return (
        <>
          {groupComponents.map(renderLayoutComponent)}
        </>
      );
    }

    return (
      <GroupContainer
        container={layoutGroup}
        id={layoutGroup.id}
        key={layoutGroup.id}
        components={groupComponents}
      />
    );
  }

  function renderLayoutComponent(layoutComponent: ILayoutComponent | ILayoutGroup) {
    if (layoutComponent.type && layoutComponent.type.toLowerCase() === 'group') {
      return RenderLayoutGroup(layoutComponent as ILayoutGroup);
    }

    const component: ILayoutComponent = layout.find((c) => c.id === layoutComponent.id) as ILayoutComponent;
    return (
      <RenderGenericComponent key={layoutComponent.id} {...component} />
    );
  }

  return (
    <Grid container={true}>
      {currentView === currentLayout && filteredLayout && filteredLayout.map(renderLayoutComponent)}
    </Grid>
  );
}

export default Form;
