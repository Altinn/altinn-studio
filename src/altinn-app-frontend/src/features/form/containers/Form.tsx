/* eslint-disable no-undef */
import Grid from '@material-ui/core/Grid';
import * as React from 'react';
import { useSelector } from 'react-redux';
import { SummaryComponent } from 'src/components/summary/SummaryComponent';
import { IRuntimeState } from '../../../types';
import { ILayout, ILayoutComponent, ILayoutGroup } from '../layout';
import { GroupContainer } from './GroupContainer';
import { renderGenericComponent } from '../../../utils/layout';
import { DisplayGroupContainer } from './DisplayGroupContainer';

export function renderLayoutComponent(layoutComponent: ILayoutComponent | ILayoutGroup, layout: ILayout) {
  switch (layoutComponent.type) {
    case 'group':
    case 'Group': {
      return RenderLayoutGroup(layoutComponent as ILayoutGroup, layout);
    }
    case 'Summary': {
      return (
        <SummaryComponent key={layoutComponent.id} {...(layoutComponent as ILayoutComponent)} />
      );
    }
    default: {
      return (
        <RenderGenericComponent key={layoutComponent.id} {...(layoutComponent as ILayoutComponent)} />
      );
    }
  }
}

function RenderGenericComponent(component: ILayoutComponent, layout: ILayout) {
  return renderGenericComponent(component, layout);
}

function RenderLayoutGroup(layoutGroup: ILayoutGroup, layout: ILayout): JSX.Element {
  const groupComponents = layoutGroup.children.map((child) => {
    let childId = child;
    if (layoutGroup.edit?.multiPage) {
      childId = child.split(':')[1] || child;
    }
    return layout.find((c) => c.id === childId) as ILayoutComponent;
  });
  const repeating = layoutGroup.maxCount > 1;
  if (!repeating) {
    // If not repeating, treat as regular components
    return (
      <DisplayGroupContainer
        container={layoutGroup}
        components={groupComponents}
        renderLayoutComponent={renderLayoutComponent}
      />
    );
  }

  return (
    <GroupContainer
      container={layoutGroup}
      id={layoutGroup.id}
      key={layoutGroup.id}
      components={groupComponents}
    />
  );
}

export function Form() {
  const [filteredLayout, setFilteredLayout] = React.useState<any[]>([]);
  const [currentLayout, setCurrentLayout] = React.useState<string>();

  const currentView = useSelector((state: IRuntimeState) => state.formLayout.uiConfig.currentView);
  const layout: ILayout =
    useSelector((state: IRuntimeState) => state.formLayout.layouts[state.formLayout.uiConfig.currentView]);

  React.useEffect(() => {
    setCurrentLayout(currentView);
  }, [currentView]);

  React.useEffect(() => {
    let componentsToRender: any[] = layout;
    let renderedInGroup: string[] = [];
    if (layout) {
      const groupComponents = layout.filter((component) => component.type.toLowerCase() === 'group');
      groupComponents.forEach((component: ILayoutGroup) => {
        let childList = component.children;
        if (component.edit?.multiPage) {
          childList = component.children.map((childId) => childId.split(':')[1] || childId);
        }
        renderedInGroup = renderedInGroup.concat(childList);
      });
    }

    if (layout) {
      componentsToRender = layout.filter((component) => {
        return !renderedInGroup.includes(component.id);
      });
    }
    setFilteredLayout(componentsToRender);
  }, [layout]);

  return (
    <Grid
      container={true}
      spacing={3}
      alignItems='flex-start'
    >
      {currentView === currentLayout && filteredLayout && filteredLayout.map((component) => {
        return renderLayoutComponent(component, layout);
      })}
    </Grid>
  );
}

export default Form;
