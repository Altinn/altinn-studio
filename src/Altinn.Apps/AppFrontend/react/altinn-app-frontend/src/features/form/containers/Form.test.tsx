import * as React from 'react';
import { Provider } from 'react-redux';
import * as renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import {
  getInitialStateMock,
  getFormLayoutStateMock,
} from '../../../../__mocks__/mocks';
import type { ILayoutState } from '../layout/formLayoutSlice';
import type { ILayoutGroup } from '../layout';

import { Form } from './Form';

describe('>>> features/form/components/Form.tsx', () => {
  let mockStore: any;
  let mockLayout: ILayoutState;
  let mockComponents: any;
  let mockGroupId: string;

  beforeAll(() => {
    window.matchMedia = jest.fn().mockImplementation((query) => {
      return {
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
      };
    });
  });

  beforeEach(() => {
    const createStore = configureStore();
    mockGroupId = 'testGroupId';
    mockComponents = [
      {
        id: 'field1',
        type: 'Input',
        dataModelBindings: {
          simple: 'Group.prop1',
        },
        textResourceBindings: {
          title: 'Title',
        },
        readOnly: false,
        required: false,
        disabled: false,
      },
      {
        id: 'field2',
        type: 'Input',
        dataModelBindings: {
          simple: 'Group.prop2',
        },
        textResourceBindings: {
          title: 'Title',
        },
        readOnly: false,
        required: false,
        disabled: false,
      },
      {
        id: 'field3',
        type: 'Input',
        dataModelBindings: {
          simple: 'Group.prop3',
        },
        textResourceBindings: {
          title: 'Title',
        },
        readOnly: false,
        required: false,
        disabled: false,
      },
    ];

    mockLayout = getFormLayoutStateMock({
      layouts: {
        FormLayout: [
          {
            id: mockGroupId,
            type: 'group',
            dataModelBindings: {
              group: 'Group',
            },
            maxCount: 3,
            children: ['field1', 'field2', 'field3'],
          } as ILayoutGroup,
        ].concat(mockComponents),
      },
    });

    const initialState = getInitialStateMock({ formLayout: mockLayout });
    mockStore = createStore(initialState);
  });

  it('+++ should match snapshot', () => {
    const rendered = renderer.create(
      <Provider store={mockStore}>
        <Form />
      </Provider>,
    );
    expect(rendered).toMatchSnapshot();
  });
});
