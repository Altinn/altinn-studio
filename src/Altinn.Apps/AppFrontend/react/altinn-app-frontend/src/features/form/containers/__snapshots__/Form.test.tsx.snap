// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> features/form/components/Form.tsx +++ should match snapshot 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-align-items-xs-flex-start"
/>
`;
